### PR TITLE
updated Profile Event 

### PR DESCRIPTION
**Pull request by iltae95**

### CODE REVIEW

**Comment by mmcgr** [comment]
I think there is a problem with printing extra timestamps, e.g.
```
new.log:@runtime;;20-03-2018 12:05:58:4,20-03-2018 12:06:00:567,2.56278
```
instead of
```
old.log:@runtime;10.3657
```
That then confuses souffle-profile which expects a double after the semicolon and gets a blank between semicolons instead.

**Comment by iltae95** [comment]
Hi Martin,

Thanks for that, it can be easily fixed. In the print methods in each of
the event classes (ProfileQuantityEvent, ProfileTimingEvent,
ProfileUtilisationEvent, ProfileMemoryEvent) there is a line,
    os << ProfileKeySingleton::instance().getText(getKey()) << ";";
at lines158, 182, 204 and 229. I think it's happening because the getText()
returns a string that already has a semicolon at the end. So, if I remove
where it adds the semi-colon in that line then this issue will be resolved.
This would probably fix the "stod" issue, right? I ask because, with that
issue, the function I thought was causing the issue has nothing to do with
stod(parsing a string as a double/float).
Also, Do you want me to correct it on my end? I don't know the protocols
here.

For the profiler section, I believe I have identified where I can make the
Profiler not print the new events like memory and utilisation at,
    src/profilerlib/OutputProcessor.cpp  line: 36 row[5] =
std::shared_ptr<CellInterface>(new Cell<std::string>(r->getName()));
but I am a little bit hesitant to change any files yet, I rather do it once
I resolve the "stod" error.


Kindest regard,
Alex Park


P.S. Sorry for such a late response.





2018-03-20 1:12 GMT+11:00 mmcgr <notifications@github.com>:

**Comment by lyndonhenry** [comment]
Nice stuff, I like it.

As described [here](https://github.com/souffle-lang/souffle/wiki/Internals#log-file-format), the profiler currently uses the following log format:



Does this PR change the existing format? 

I ask specifically because I have a use case requiring conversion of this format to CSV (as introduced in #615), and a new format may invalidate this conversion. 

If such an invalidation does occur, this should have been picked up by the unit test for the format converter and not got past Travis, but unfortunately I am rather awful at writing unit tests (sorry).

I can easily update the converter for a new format, that is no trouble at all. However the new format should be documented on the wiki so I, and others in future, know what it is. Please include enough information that the type of data in each column is made obvious.

On a bit of a side note, if this does change the format, the profiler should be pretty comprehensively tested for corner and edge cases (e.g. escaped quoted, quoted delimiters, escaped escapes, etc) to ensure that nothing is broken -- again, the tests should pick this up, but due to the "bug" described in issue #490, they do not. I don't really care about this as much as the formatting at the moment, but improved profiling seems worth less than having a stable profiler that works properly.

**Comment by mmcgr** [comment]
The stod error happens because of the ;; in the runtime row. What's between the semicolons is passed to stod, which then fails. I would expect the same problem for all the other lines, though, since they have a date where souffle-profile expects a double.

For now, disable output of the new information, and output the old information in the old format, i.e., without the start and end timestamps. That should be sufficient to not crash, and pass the tests.

When you've made those changes just push them to github and this PR will update automatically.

**Comment by iltae95** [comment]
I made and pushed the changes and checked that it passed all the tests on
Travis.
I edited dump method in ProfileEvent.h so that it skips memory and
utilisation events.

2018-03-21 11:00 GMT+11:00 mmcgr <notifications@github.com>:

**Comment by mmcgr** [comment]
Something went wrong with github/travis and the PR was not being built. I've added a tiny bit of refactoring just to trigger a new build

File: src/ProfileEvent.h [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+    /** Convert ProfileTimePoint(time) to string */
+    const std::string timeToString(ProfileTimePoint& time) {
+        std::chrono::milliseconds ms =
+                std::chrono::duration_cast<std::chrono::milliseconds>(time.time_since_epoch());
+
+        std::chrono::seconds s = std::chrono::duration_cast<std::chrono::seconds>(ms);
+        std::time_t t = s.count();
+        std::size_t fractional_seconds = ms.count() % 1000;
+
+        char buffer[80];
```
Comment (mmcgr) [inline src/ProfileEvent.h]: You could do this with C++ tools. Something like
```
std::stringstream ss;
ss << std::put_time(std::localtime(&time), "%F %T") << ":" << fractional_seconds;
return ss.str();
```
This can wait for a later PR if needed.

File: src/ProfileEvent.h [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+    }
+};
+/**
+ * Profile Utilisation Event
+ */
+class ProfileUtilisationEvent : public ProfileEvent {
+private:
+    /** resouce statistics */
+    struct rusage ru;
+    // static unsigned long long lastTotalUser, lastTotalUserLow, lastTotalSys, lastTotalIdle;
```
Comment (mmcgr) [inline src/ProfileEvent.h]: Remove commented out code

File: src/ProfileEvent.h [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+    }
+
+    /** Print event */
+    void print(std::ostream& os) override {
+        os << ProfileKeySingleton::instance().getText(getKey());
+        /* system CPU time used */
+        double t = (double)ru.ru_stime.tv_sec * 1000000.0 + (double)ru.ru_stime.tv_usec;
+        os << getStartString() << ",";
+        os << std::to_string(t);
+        // os << "," << getCurrentValue();
```
Comment (mmcgr) [inline src/ProfileEvent.h]: Remove commented out code

File: src/ProfileEvent.h [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+class ProfileEventSingleton {
+private:
+    /**  Profile Timer */
+    class ProfileTimer {
+    private:
+        /** time interval between per utilisation read */
+        Interval t;
+
+        /** timer is running */
+        bool running;
```
Comment (mmcgr) [inline src/ProfileEvent.h]: Initialise running as false here, rather than the constructor

File: src/ProfileEvent.h [hunk 5]
```diff
@@ -0,0 +1,358 @@
...
+        }
+    }
+    static ProfileEventSingleton& instance() {
+        static ProfileEventSingleton singleton;
+        return singleton;
+    }
+
+    /** Make new time event */
+    ProfileTimingEvent* makeTimingEvent(const std::string& txt) {
+        ProfileTimingEvent* e = new ProfileTimingEvent(txt);
```
Comment (mmcgr) [inline src/ProfileEvent.h]: Save a little time for the reader and use `auto`, since you name the class twice on one line.

File: src/ProfileEvent.h [hunk 6]
```diff
@@ -0,0 +1,358 @@
...
+        os << number;
+        os << std::endl;
+    }
+};
+/**
+ * Profile Utilisation Event
+ */
+class ProfileUtilisationEvent : public ProfileEvent {
+private:
+    /** resouce statistics */
```
Comment (mmcgr) [inline src/ProfileEvent.h]: resource

File: src/ProfileEvent.h [hunk 7]
```diff
@@ -0,0 +1,358 @@
...
+#include <ctime>
+#include <functional>
+#include <iostream>
+#include <list>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <vector>
+#include <assert.h>
```
[the remark this answers is not in the document]
Comment (mmcgr) [reply src/ProfileEvent.h]: The auto-formatting broke the comment. It was meant to say `Use #include <cassert>`

**Comment by mmcgr** [comment]
Travis package tests were failing due to the old g++ version used. I changed the code back from what I suggested so it could be built. A more elegant solution will wait until the packaging step is updated.
